src/Features/CSharp/Portable/UseExpressionBodyForLambda/UseExpressionBodyForLambdaCodeStyleProvider.cs

```diff
@@ -0,0 +1,434 @@
+            }
+
+            var canOffer = expressionBodyOpt.TryConvertToStatement(
+                semicolonTokenOpt: default, createReturnStatementForExpression: false, out _);
+            if (!canOffer)
+            {
+                // Couldn't even convert the expression into statement form.
+                return false;
+            }
+
+            var languageVersion = ((CSharpParseOptions)declaration.SyntaxTree.Options).LanguageVersion;
+            if (expressionBodyOpt.IsKind(SyntaxKind.ThrowExpression) &&
+                languageVersion < LanguageVersion.CSharp7)
+            {
+                // Can't convert this prior to C# 7 because ```a => throw ...``` isn't allowed.
+                return false;
+            }
+
+            return true;
+        }
+
+        private static Location GetDiagnosticLocation(LambdaExpressionSyntax declaration)
+            => Location.Create(declaration.SyntaxTree,
+                    TextSpan.FromBounds(declaration.SpanStart, declaration.ArrowToken.Span.End));
+
+        #endregion
+
+        #region fixing
+
+        protected override Task<ImmutableArray<CodeAction>> ComputeCodeActionsAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var codeAction = new MyCodeAction(
+                diagnostic.GetMessage() + "-Fix",
+                c => FixWithSyntaxEditorAsync(document, diagnostic, c));
+
+            return Task.FromResult(ImmutableArray.Create<CodeAction>(codeAction));
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var diagnostic in diagnostics)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                AddEdits(editor, semanticModel, diagnostic, cancellationToken);
+            }
+        }
+
+        private void AddEdits(
+            SyntaxEditor editor, SemanticModel semanticModel,
+            Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var declarationLocation = diagnostic.AdditionalLocations[0];
+            var declaration = (LambdaExpressionSyntax)declarationLocation.FindNode(getInnermostNodeForTie: true, cancellationToken);
+            var useExpressionBody = diagnostic.Properties.ContainsKey(nameof(UseExpressionBody));
+
+            editor.ReplaceNode(
+                declaration,
+                Update(semanticModel, declaration, useExpressionBody));
+        }
+
+        #endregion
+
+        #region refactoring
+
+        protected override async Task<ImmutableArray<CodeAction>> ComputeReverseRefactoringsAsync(
+            Document document, TextSpan span, ExpressionBodyPreference option, CancellationToken cancellationToken)
```

Comment:
Note how this feature is required to think about what the right experience is here.  Effectively, this is saying "the user has set some expression body preference, and the analyzer portion of the code will enforce this setting.  however, we should still give them a refactoring to go in the reverse direction."